File: min_window_substring/min_window_substring.py
from collections import Counter

def min_window_substring(s, t):
    if not t or not s:
        return ""
    
    need = Counter(t)
    need_count = len(need)
    window= {}
    have = 0
    res = ""
    res_length = float('inf')
    L = 0
    
    for R in range(len(s)):
        c = s[R]
        window[c] = window.get(c, 0) + 1
        if c in need and window.get(c) == need.get(c):
            have += 1
        print("for loop: ", window, have)
        while have == need_count:
            print("While loop: ", window)
            if(R - L +1) < res_length:
                res_length = R - L + 1
                res = s[L: R + 1]
            print("res_len", res_length)
            left_c = s[L]
            window[left_c] -= 1
            print(window[left_c], need[left_c])
            if left_c in need and window[left_c] < need[left_c]:
                have -= 1
            L += 1
    return res
    
    print(window)

s = "ADOBECODEBANC"
t = "ABC"
print("The shortest substring is: ", min_window_substring(s, t));

File: min_window_substring/min_window_substring.ts
export {};

function minWindowSubstring(s: string, t: string): string {
  if (t.length === 0 || s.length === 0) return "";

  const need = new Map<string, number>();
  for (const c of t) need.set(c, (need.get(c) ?? 0) + 1);
  const needCount = need.size;

  const window = new Map<string, number>();
  let have = 0;
  let res = "";
  let resLen = Infinity;
  let L = 0;

  for (let R = 0; R < s.length; R++) {
    const c = s[R];
    window.set(c, (window.get(c) ?? 0) + 1);
    if (need.has(c) && window.get(c) === need.get(c)) have++;
    while (have === needCount) {
      if (R - L + 1 < resLen) {
        resLen = R - L + 1;
        res = s.slice(L, R + 1);
      }
      const leftC = s[L];
      window.set(leftC, (window.get(leftC) ?? 0) - 1);
      if (need.has(leftC) && (window.get(leftC) ?? 0) < (need.get(leftC) ?? 0)) have--;
      L++;
    }
  }

  return res;
}

let s = "ADOBECODEBANC";
let t = "ABC";
console.log("The shortest substring is: ", minWindowSubstring(s, t));
